text/0000-name-resolution.md

```diff
@@ -0,0 +1,420 @@
+}
+
+mod bar {
+    pub trait Qux;
+}
+
+mod boz {
+    use foo::*;
+    use bar::Qux; // Shadows only in the type name space.
+
+    fn f(x: &Qux) {   // bound to bar::Qux.
+        let _ = Qux;  // bound to foo::Qux.
+    }
+}
+```
+
+This change is discussed in [issue 31337](https://github.com/rust-lang/rust/issues/31337).
+
+
+## Changes to the implementation
+
+Note: below I talk about "the binding table", this is sort of hand-waving. I'm
+envisaging a sets-of-scopes system where there is effectively a single, global
+binding table. However, the details of that are beyond the scope of this RFC.
+One can imagine "the binding table" means one binding table per scope, as in the
+current system.
+
+Currently, parsing and macro expansion happen in the same phase. With this
+proposal, we add import resolution to that mix too. Binding tables as well as
+the AST will be produced by libsyntax. Name lookup will continue to be done
+where name resolution currently takes place.
+
+To resolve imports, the algorithm proceeds as follows: we start by parsing as
+much of the program as we can; like today we don't parse macros. When we find
+items which bind a name, we add the name to the binding table. When we find an
+import which can't be resolved, we add it to a work list. When we find a glob
+import, we have to record a 'back link', so that when a public name is added for
+the supplying module, we can add it for the importing module.
+
+We then loop over the work list and try to lookup names. If a name has exactly
+one best binding then we use it (and record the binding on a list of resolved
+names). If there are zero, or more than one possible binding, then we put it
+back on the work list. When we reach a fixed point, i.e., the work list no
+longer changes, then we are done. If the work list is empty, then
+expansion/import resolution succeeded, otherwise there are names not found, or
+ambiguous names, and we failed.
+
+As we are looking up names, we record the resolutions in the binding table. If
+the name we are looking up is for a glob import, we add bindings for every
+accessible name currently known.
+
+To expand a macro use, we try to resolve the macro's name. If that fails, we put
+it on the work list. Otherwise, we expand that macro by parsing the arguments,
+pattern matching, and doing hygienic expansion. We then parse the generated code
+in the same way as we parsed the original program. We add new names to the
+binding table, and expand any new macro uses.
+
+If we add names for a module which has back links, we must follow them and add
+these names to the importing module (if they are accessible). When following
+these back links, we check for cycles, signaling an error if one is found.
+
+In pseudo-code:
+
+```
+// Assumes parsing is already done, but the two things could be done in the same
+// pass.
+fn parse_expand_and_resolve() {
+    loop until fixed point {
+        loop until fixed point {
+            process_names()
```

Comment:
I think `process_names()` should be called just before the inner fix-point loop, not in it.